watcher.go

```diff
@@ -185,75 +185,88 @@ func (w *Watcher) WaitCh(ctx context.Context) <-chan error {
 	return errCh
 }
 
-// Wait blocks until new a watched value changes or until context is closed
-// or exceeds its deadline.
-func (w *Watcher) Wait(ctx context.Context) error {
-	w.stopCh.drain() // in case Stop was already called
+type notifierMap map[string]struct{}
 
+func (w *Watcher) wait(ctx context.Context) (notifierMap, error) {
+	empty := struct{}{}
 	// send waiting notification, only used for testing
 	select {
-	case w.waitingCh <- struct{}{}:
+	case w.waitingCh <- empty:
 	default:
 	}
 
-	// combine cache and changed updates so we don't forget one
-	dataUpdate := func(v *view) (notify bool) {
+	dataUpdate := func(v *view, notifiers notifierMap) notifierMap {
 		id := v.ID()
 		w.cache.Save(id, v.Data())
 		for _, n := range w.tracker.notifiersFor(v) {
 			if n.Notify(v.Data()) && !w.Buffering(n) {
-				notify = true
+				notifiers[n.ID()] = empty
 			}
 		}
-		return notify
+		return notifiers
 	}
-	for {
-		select {
-		case view := <-w.dataCh:
-			// this case (<-w.dataCh) only happens if there is new/udpated data
-			notify := dataUpdate(view)
-			// Drain all dependency data. Prevents re-rendering templates over
-			// and over when a large batch of dependencies are updated.
-			// See consul-template GH-168 for background.
-			for drain := true; drain; {
-				select {
-				case view := <-w.dataCh:
-					if dataUpdate(view) && !notify {
-						notify = true
-					}
-				case <-time.After(time.Microsecond):
-					drain = false
-				}
-			}
-			if notify {
-				return nil
+
+	// use map to dedup results
+	notifiers := make(notifierMap)
+
+	select {
+	case v := <-w.dataCh:
+		notifiers = dataUpdate(v, notifiers)
+		// this case (<-w.dataCh) only happens if there is new/udpated data
+		for drain := true; drain; {
+			select {
+			case v := <-w.dataCh:
+				notifiers = dataUpdate(v, notifiers)
+			case <-time.After(time.Microsecond):
+				drain = false
 			}
-		case <-w.bufferTrigger:
-			// A template is now ready to be rendered, though there might be a
-			// few ready around the same time if they have the same dependencies.
-			// Drain the channel similar for the dataCh above.
-			for {
-				select {
-				case <-w.bufferTrigger:
-				case <-time.After(time.Microsecond):
-					return nil
-				}
+		}
+		return notifiers, nil
+	case nID := <-w.bufferTrigger:
+		notifiers[nID] = empty
+		// A template is now ready to be rendered, though there might be a
+		// few ready around the same time if they have the same dependencies.
+		// Drain the channel similar for the dataCh above.
+		for drain := true; drain; {
+			select {
+			case nID := <-w.bufferTrigger:
+				notifiers[nID] = empty
+			case <-time.After(time.Microsecond):
+				drain = false
 			}
+		}
+		return notifiers, nil
+	case <-w.stopCh:
+		return nil, StopError
 
-		case <-w.stopCh:
-			return nil
+	case err := <-w.errCh:
+		// Push the error back up the stack
+		return nil, err
+
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+var StopError = fmt.Errorf("Stop")
```

Comment:
I find that it's easiest to find global constants and variables when they're grouped together at the beginning of the file or a separate err file altogether. Could you move this up?

Also a nit [idiomatic go suggestion](https://pkg.go.dev/net/http#pkg-variables) for error naming: `ErrXXX` for error variables and `XXXError` for error types.

Reply:
Oh found a better example of error naming https://talks.golang.org/2014/names.slide#14

Reply:
Updated the name to ErrStop (also updated RegistryErr to ErrRegistry) and moved it to the top next to ErrRegistry.